### PR TITLE
C++ migration: add a C++ implementation of `FSTExponentialBackoff`

### PR DESCRIPTION
**Pull request by var-const**
This is a pretty close port of `FSTExponentialBackoff`. The changes are pretty minor:
* delay is calculated using `<chrono>` duration types, not plain numbers, which should be a little more type-safe;
* split a piece of code into a `ClampDelay` function, because it's reasonably close to [std::clamp](https://en.cppreference.com/w/cpp/algorithm/clamp);
* rephrased the class-level comment to make it clearer that the first attempt always has delay = 0;
* added simple tests (other platforms don't have tests for this).

Also make sure that canceling a `DelayedOperation` is always valid.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  delayed_operation_ = queue_->EnqueueAfterDelay(delay_with_jitter, timer_id_,
+                                                 std::move(operation));
+
+  // Apply backoff factor to determine next delay, but ensure it is within
+  // bounds.
+  current_base_ = ClampDelay(current_base_ * backoff_factor_);
+}
+
+/** Returns a random value in the range [-current_base_/2, current_base_/2] */
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: I'd move this doc string to the header (since it's discussing what the method does).

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  util::AsyncQueue* queue_;
+  util::TimerId timer_id_;
+  util::DelayedOperation delayed_operation_;
+
+  double backoff_factor_;
+  RealSeconds current_base_{0};
+  util::AsyncQueue::Milliseconds initial_delay_;
+  util::AsyncQueue::Milliseconds max_delay_;
+  util::SecureRandom secure_random_;
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: A *secure* rng is probably overkill for this and could cause the system to loose entropy for operations that truly require one. Consider switching to a pseudo rng.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+auto ExponentialBackoff::GetDelayWithJitter() -> RealSeconds {
+  std::uniform_real_distribution<double> distribution;
+  const auto random_double = distribution(secure_random_);
+  return (random_double - 0.5) * current_base_;
+}
+
+auto ExponentialBackoff::ClampDelay(const RealSeconds delay) const
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: (as above)

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    -> RealSeconds {
+  if (delay < initial_delay_) {
+    return initial_delay_;
+  }
+  if (delay > max_delay_) {
+    return max_delay_;
+  }
+  return delay;
+}
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: This looks pretty close to std::clamp. Any reason not to just use that directly?

File: Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#include <chrono>  // NOLINT(build/c++11)
+
+#include "Firestore/core/src/firebase/firestore/remote/exponential_backoff.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/executor_std.h"
+#include "Firestore/core/test/firebase/firestore/util/async_tests_util.h"
+#include "absl/memory/memory.h"
+#include "gtest/gtest.h"
+
+using firebase::firestore::util::AsyncQueue;
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc]: I recently read http://go/totw/119 which suggests moving these using declarations inside the namespace statements. We're not at all consistent with this though. (If anything, we more consistently don't do this.)

Optional.

File: Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    backoff.BackoffAndRun([&] { signal_finished(); });
+  });
+
+  // The chosen value of initial_delay is large enough that it shouldn't be
+  // realistically possible for backoff to finish already.
+  queue.RunScheduledOperationsUntil(timer_id);
+  EXPECT_TRUE(WaitForTestToFinish());
+}
+
+}  // namespace remote
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc]: There's a number of missing tests, but since the other ports have none... shrug. Since most of the missing tests involve interactions with minimum/maximum delays, they'd be ... somewhat less than ideal ... without a way to control the clock. Deferring until that's figured out seems fine to me.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: On all the platforms we care about `SecureRandom` does the right thing, namely that it draws from the entropy pool only to seed and reseed a PRNG but otherwise does not draw directly from system entropy.

So while SecureRandom may be overkill, it's not harmful, and since it already exists, it's reasonable to continue using it here.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+   *     could be as much as `1.5*max_delay`.
+   */
+  ExponentialBackoff(util::AsyncQueue* queue,
+                     util::TimerId timer_id,
+                     double backoff_factor,
+                     util::AsyncQueue::Milliseconds initial_delay,
+                     util::AsyncQueue::Milliseconds max_delay)
+      : queue_{queue},
+        timer_id_{timer_id},
+        backoff_factor_{backoff_factor},
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: Done.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  util::AsyncQueue* queue_;
+  util::TimerId timer_id_;
+  util::DelayedOperation delayed_operation_;
+
+  double backoff_factor_;
+  RealSeconds current_base_{0};
+  util::AsyncQueue::Milliseconds initial_delay_;
+  util::AsyncQueue::Milliseconds max_delay_;
+  util::SecureRandom secure_random_;
+};
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: Done.

I'm always a little wary about making data members `const` because it makes the class non-reassignable, and if the caller wants the instance to be `const`, they may simply declare it so. OTOH, this class looks fine being non-reassignable.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  /**
+   * Resets the backoff to the maximum delay (e.g. for use after
+   * a RESOURCE_EXHAUSTED error).
+   */
+  void ResetToMax() {
+    current_base_ = max_delay_;
+  }
+
+  /**
+   * Waits for `current_base` seconds, increases the delay and runs the
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: Done.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  delayed_operation_ = queue_->EnqueueAfterDelay(delay_with_jitter, timer_id_,
+                                                 std::move(operation));
+
+  // Apply backoff factor to determine next delay, but ensure it is within
+  // bounds.
+  current_base_ = ClampDelay(current_base_ * backoff_factor_);
+}
+
+/** Returns a random value in the range [-current_base_/2, current_base_/2] */
+auto ExponentialBackoff::GetDelayWithJitter() -> RealSeconds {
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: Done.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc (hunk 1 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: Done.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: @rsgowman It's a straightforward copy from the Objective-C++ code.

@wilhuff Thanks for the explanation!

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc (hunk 3 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: Done.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc (hunk 4 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: `std::clamp` is C++17, unfortunately, and I don't think it's in Abseil, weirdly.

File: Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc (hunk 5 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc]: Done.

I think adding a leading `::` is a bit overkill, though. What do you think?

Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/remote/exponential_backoff_test.cc]: For sibling namespaces, that link suggests that fully qualifying (with a leading `::`) is always correct, but does allow for relative usage as well. Looking at this specific case, it seems unlikely that we'd introduce any of these namespaces:

```
::firebase::firebase
::firebase::firestore::firebase
::firebase::firestore::remote::firebase
```
(and if we ever do, we've got bigger problems).

So I think dropping the `::` would be fine. OTOH, including it would allow us to never have to stop and think about it.

FWIW, after I read that link, I starting moving my `using` declarations inside of my namespaces, but I didn't add leading `::`s. Probably worth kicking this over to the chat and quickly deciding as a group what we want to do. (Though we've already spent more time discussing it in this PR than it's worth. This is a pretty minor issue.)

For the purposes of this PR, I'd recommend doing whatever you like. If we decide to "fix" it, we can do so all at once later.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+   */
+  void BackoffAndRun(util::AsyncQueue::Operation&& operation);
+
+  /** Cancels any pending backoff operation scheduled via `BackoffAndRun`. */
+  void Cancel() {
+    delayed_operation_.Cancel();
+  }
+
+ private:
+  using RealSeconds = std::chrono::duration<double>;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: Why mix units like this?

The API takes milliseconds and uses milliseconds to call the async queue.

The other ports e.g. TS just use milliseconds throughout.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.h [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+   *     delay after each attempt.
+   * @param initial_delay The initial delay (used as the base delay on the first
+   *     retry attempt, that is, the second attempt). Note that jitter will
+   *     still be applied, so the actual delay could be as little as
+   *     `0.5*initial_delay`.
+   * @param max_delay The maximum base delay after which no further backoff is
+   *     performed. Note that jitter will still be applied, so the actual delay
+   *     could be as much as `1.5*max_delay`.
+   */
+  ExponentialBackoff(util::AsyncQueue* queue,
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/exponential_backoff.h]: In the other ports this comment is mostly the class comment.

**Comment by var-const** [comment]
> Please remember to apply local changes (e.g. clamp) to the other platforms so that they remain comparable.

Will do.

File: Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+  // Apply backoff factor to determine next delay, but ensure it is within
+  // bounds.
+  current_base_ = ClampDelay(
+      chr::duration_cast<Milliseconds>(current_base_ * backoff_factor_));
+}
+
+ExponentialBackoff::Milliseconds ExponentialBackoff::GetDelayWithJitter() {
+  std::uniform_real_distribution<double> distribution;
+  const auto random_double = distribution(secure_random_);
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: Also see line 58 above:
`const auto delay_with_jitter = current_base_ + GetDelayWithJitter();`

Now that the duration_cast is gone, I no longer know what the type is.

[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: A(n unofficial) dissenting opinion: go/totw/labs/const#where
"Local variables that do not change."

Also see "What's the point of using const on pass-by-value parameters?" in that same doc.

(FWIW, I like const on member variables, but have generally avoided it on instance variables due to the extra noise. After reading that doc, I might try it a bit on some personal stuff to see how much I [dis]like it. For the moment, I have no opinion on this topic so am happy going either way here.)

[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/exponential_backoff.cc]: Google's internal Java recommendations on `final` go the other way: go/java-practices/final#params_locals.

I recognize that each of these languages is different but the trade-off seems essentially the same.
